TestingStrategy.md

```diff
@@ -0,0 +1,195 @@
+
+export const Default: Story = {
+  args: {
+    buttonText: 'Button'
+  }
+};
+```
+
+## Jest
+
+Jest unit tests should cover stand alone functions that cannot be covered by either storybook or Cypress. These tests cover more complicated business logic that would be cumbersome to test in Cypress component tests.
+
+For example: we don't need to test the UI for every use-case of a dice roller. The UI only needs to accept a /roll command and spit out a number we expect. The business logic for that dice roller however should be extensively covered by a unit test.
+
+**Example Cypress Test**
+
+```typescript
+it('should roll a d20', () => {
+  cy
+    .stub(Math, 'random')
+    .returns(0.5);
+
+  cy
+    .getToolbarFooter()
+    .find('input')
+    .eq(0)
+    .type('/roll 1d20+4')
+    .type('{enter}');
+
+  cy
+    .getToolbarFooter()
+    .find('input')
+    .eq(0)
+    .invoke('attr', 'value')
+    .should('equal', '');
+  
+  cy
+    .get('[data-test-id="toolbar-footer-rolled-value"]')
+    .should('contain', '15');
+});
+```
+
+**Example Jest Test**
+
+```typescript
+it('should handle a /roll command', () => { ... });
+
+it('should handle a /roll command with a minus', () => { ... });
+
+it('should handle a /roll command with no addition', () => { ... });
+
+it('should handle a malformed command', () => { ... });
+
+it('should handle an absent command', () => { ... });
+
+// more edge cases
+```
+
+## Cypress
+
+Cypress component tests should accompany each page in the app. Specifically we use component tests vs e2e tests to avoid the problem of needed to mock database environments and to simplify execution, for what is basically a hobby project.
+
+- We can largely trust that React Router and paths work for this project.
+- The backend has its own suite of tests and as long as the contracts remain the same between frontend and backend, that should be enough for most of our use cases.
+- Component tests largely allow us to ignore the need for rolling back transactions, managing multiple db environments and infrastructure, and database cleanup.
+
+### Cypress Commands
+
+Cypress allows for the creation of helper functions to make tests more readable when querying the dom. We should only make these functions for components that we expect to reuse across multiple tests.
+
+These commands should also be simple and single-use. We want to avoid opinionated helper functions so we don't end up with 4 iterations of:
+
+- getButton()
+- getButtonByIndex()
+- getButtonAndClick()
+- getSecondButtonAndClick()
+
+The test should be deciding how to best orchestrate these helper functions instead of shoving test functionality in the helpers.
+
+For example, this is a simple command for retrieving a button and its use in a test.
+
+**/cypress/support/commands.ts**
+
+```typescript
+const getButton = () => {
+  return cy.get('[data-test-id="dm-screen-design-system-button"]');
+};
+
+type ChainableElement = Cypress.Chainable<JQuery<HTMLElement>>;
```

Comment:
Not a problem, just surprise. jQuery? O: neat

Reply:
Haha, yup that was a surprise to me too. It took a bit of hunting to get that typing right.